File: app/core/ratings.py
"""Ratings extraction from Plex/Jellyfin metadata."""
from __future__ import annotations

import logging
from typing import Any, Optional

log = logging.getLogger(__name__)


def _safe_float(x: Any) -> Optional[float]:
    """Parse ``x`` to a float, returning None for missing/unparseable input.

    Rating fields in Plex/Jellyfin are optional and sometimes arrive as
    strings, None, or obviously-invalid values. Silent failure is intentional
    here -- these are informational; a missing rating should not raise.
    """
    if x is None:
        return None
    try:
        return float(x)
    except (ValueError, TypeError):
        return None


def extract_plex_ratings(item: dict) -> dict:
    """Extract ratings from a Plex metadata item.

    Plex stores:
    - `rating` → Rotten Tomatoes critic score (0-10 scale, multiply by 10 for %)
    - `audienceRating` → Community/IMDb-style rating (0-10 scale)
    - `contentRating` → Age rating (PG-13, R, etc.) - not a quality score

    Returns {"critic_rating": int|None, "audience_rating": float|None}.
    critic_rating is RT percentage (0-100), audience_rating is 0-10 scale.
    """
    result: dict = {"critic_rating": None, "audience_rating": None}

    # Plex's `rating` is the critic score (RT) on a 0-10 scale
    plex_rating = _safe_float(item.get("rating"))
    if plex_rating is not None:
        result["critic_rating"] = round(plex_rating * 10)  # Convert to 0-100%

    # Plex's `audienceRating` is the audience/community score on a 0-10 scale
    audience = _safe_float(item.get("audienceRating"))
    if audience is not None:
        result["audience_rating"] = round(audience, 1)

    return result


def extract_jellyfin_ratings(item: dict) -> dict:
    """Extract ratings from a Jellyfin metadata item.

    Jellyfin stores:
    - `CriticRating` → Rotten Tomatoes critic score (0-100%)
    - `CommunityRating` → Community/IMDb-style rating (0-10 scale)

    Returns {"critic_rating": int|None, "audience_rating": float|None}.
    """
    result: dict = {"critic_rating": None, "audience_rating": None}

    critic = _safe_float(item.get("CriticRating"))
    if critic is not None:
        result["critic_rating"] = int(critic)

    community = _safe_float(item.get("CommunityRating"))
    if community is not None:
        result["audience_rating"] = round(community, 1)

    return result


def extract_ratings(item: dict, source: str = "plex") -> dict:
    """Extract ratings from a media server item.

    Args:
        item: Plex or Jellyfin item dict
        source: "plex" or "jellyfin"

    Returns {"critic_rating": int|None, "audience_rating": float|None}
    """
    if source == "jellyfin":
        return extract_jellyfin_ratings(item)
    return extract_plex_ratings(item)

File: internal/services/ratings/ratings.go
// Package ratings extracts critic and audience ratings from Plex and Jellyfin
// item metadata.
package ratings

import (
	"log/slog"
	"math"

	"github.com/rinseaid/reclaimer/internal/models"
)

// safeFloat parses v to a float64, returning nil for missing or invalid values.
func safeFloat(v any) *float64 {
	if v == nil {
		return nil
	}
	switch n := v.(type) {
	case float64:
		return &n
	case float32:
		f := float64(n)
		return &f
	case int:
		f := float64(n)
		return &f
	case int64:
		f := float64(n)
		return &f
	case string:
		// Ratings should be numeric; log and skip strings.
		return nil
	}
	return nil
}

// ExtractPlexRatings extracts ratings from a Plex metadata item.
//
// Plex stores:
//   - rating: Rotten Tomatoes critic score on a 0-10 scale (multiply by 10 for %)
//   - audienceRating: community / IMDb-style rating on a 0-10 scale
func ExtractPlexRatings(item map[string]any) models.Ratings {
	var r models.Ratings

	if pf := safeFloat(item["rating"]); pf != nil {
		v := int(math.Round(*pf * 10)) // convert 0-10 -> 0-100%
		r.CriticRating = &v
	}

	if af := safeFloat(item["audienceRating"]); af != nil {
		v := math.Round(*af*10) / 10 // one decimal place
		r.AudienceRating = &v
	}

	return r
}

// ExtractJellyfinRatings extracts ratings from a Jellyfin metadata item.
//
// Jellyfin stores:
//   - CriticRating: Rotten Tomatoes critic score already on a 0-100% scale
//   - CommunityRating: community / IMDb-style rating on a 0-10 scale
func ExtractJellyfinRatings(item map[string]any) models.Ratings {
	var r models.Ratings

	if cf := safeFloat(item["CriticRating"]); cf != nil {
		v := int(*cf)
		r.CriticRating = &v
	}

	if af := safeFloat(item["CommunityRating"]); af != nil {
		v := math.Round(*af*10) / 10 // one decimal place
		r.AudienceRating = &v
	}

	return r
}

// ExtractRatings dispatches to the appropriate extractor based on source
// ("plex" or "jellyfin").
func ExtractRatings(item map[string]any, source string) models.Ratings {
	switch source {
	case "jellyfin":
		return ExtractJellyfinRatings(item)
	case "plex":
		return ExtractPlexRatings(item)
	default:
		slog.Warn("Unknown rating source, defaulting to plex", "source", source)
		return ExtractPlexRatings(item)
	}
}
